### PR TITLE
[MIG] l10n_ch_payment_slip: Migration to 11.0

### PR DESCRIPTION
**Pull request by yvaucher**
Update headers
Port code to python3
Update README
Fix tests with correct report render calls
Adapt code to report module split

TODO: 
- [x] fixing tests
- [x] v11 + tests -> moved in another module
- [x] put bvr parameters on res.confing.settings
- [ ] edit README.rst to explain the difference with Odoo official addon.

### CODE REVIEW

**Review by grindtildeath — COMMENTED** [review]
https://github.com/OCA/l10n-switzerland/pull/373/files#diff-788bc5e9535f52e4fdb14b3cf5d6a2e2R15

odoo.report doesn't exist in v11, so it crashes if submodule is in addons_path

```
2017-11-20 16:18:15,761 1 ERROR ? werkzeug: Error on request:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/werkzeug/serving.py", line 193, in run_wsgi
    execute(self.server.app)
  File "/usr/local/lib/python3.5/dist-packages/werkzeug/serving.py", line 181, in execute
    application_iter = app(environ, start_response)
  File "/opt/odoo/src/odoo/service/server.py", line 251, in app
    return self.app(e, s)
  File "/opt/odoo/src/odoo/service/wsgi_server.py", line 166, in application
    return application_unproxied(environ, start_response)
  File "/opt/odoo/src/odoo/service/wsgi_server.py", line 154, in application_unproxied
    result = handler(environ, start_response)
  File "/opt/odoo/src/odoo/http.py", line 1297, in __call__
    self.load_addons()
  File "/opt/odoo/src/odoo/http.py", line 1319, in load_addons
    m = __import__('odoo.addons.' + module)
  File "/opt/odoo/src/odoo/modules/module.py", line 82, in load_module
    exec(open(modfile, 'rb').read(), new_mod.__dict__)
  File "<string>", line 1, in <module>
    
  File "/opt/odoo/external-src/l10n-switzerland/l10n_ch_payment_slip/models/__init__.py", line 2, in <module>
    from . import payment_slip
  File "/opt/odoo/external-src/l10n-switzerland/l10n_ch_payment_slip/models/payment_slip.py", line 15, in <module>
    from odoo.report import report_sxw
ImportError: No module named 'odoo.report'
```

File: l10n_ch_payment_slip/__manifest__.py [hunk 1]
```diff
@@ -1,32 +1,28 @@
-# -*- coding: utf-8 -*-
-# © 2012-2016 Camptocamp SA
+# Copyright 2012-2016 Camptocamp SA
```
Comment (simahawk) [inline l10n_ch_payment_slip/__manifest__.py:1]: 2012-2018, same for other py files

File: l10n_ch_payment_slip/models/company.py [hunk 2]
```diff
@@ -1,62 +1,71 @@
-# -*- coding: utf-8 -*-
-# © 2012-2015 Camptocamp SA
+# Copyright 2012-2015 Camptocamp SA
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 from odoo import models, fields
 
 
 class ResCompany(models.Model):
-    """override company in order to add bvr vertical and
+    """override company in order to add ISR vertical and
```
Comment (simahawk) [inline l10n_ch_payment_slip/models/company.py]: ```"""Add ISR vertical and Horizontal print delta."""```

File: l10n_ch_payment_slip/models/invoice.py [hunk 3]
```diff
@@ -24,8 +23,8 @@ class AccountMoveLine(models.Model):
 
 
 class AccountInvoice(models.Model):
-    """Inherit account.invoice in order to add bvr
-    printing functionnalites. BVR is a Swiss payment vector"""
+    """Inherit account.invoice in order to add ISR
```
Comment (simahawk) [inline l10n_ch_payment_slip/models/invoice.py]: """Add ISR (Swiss payment vector) printing features."""

File: l10n_ch_payment_slip/report/ir_action.py [hunk 4]
```diff
@@ -1,46 +1,107 @@
...
-    def _lookup_report(self, name):
-        self.env.cr.execute(
-            "SELECT * FROM ir_act_report_xml WHERE report_name=%s", (name,))
-        report = self.env.cr.dictfetchone()
-        if report and report['report_type'] == 'reportlab-pdf':
-            return report['report_name']
+    @api.multi
+    def _generate_one_slip_per_page_from_invoice_pdf(self, report_name=None):
+        """Generate payment slip PDF(s) from report model.
+        If there is many pdf they are merged in memory or on
```
Comment (simahawk) [inline l10n_ch_payment_slip/report/ir_action.py]: blank line after 1st one.

Comment (simahawk) [inline l10n_ch_payment_slip/report/ir_action.py]: "If there are many PDF..."

**Review by simahawk — CHANGES_REQUESTED** [review]
LGTM, just nitpicking non-blocking comments ;)

File: l10n_ch_payment_slip/__manifest__.py (hunk 1 above)
Comment (yvaucher) [reply l10n_ch_payment_slip/__manifest__.py:1]: When significant changes were made in py files I updated it already (was in 2017 BTW)

File: l10n_ch_payment_slip/views/company.xml [hunk 5]
```diff
@@ -2,25 +2,25 @@
 <openerp>
     <data>
        <record model="ir.ui.view" id="company_form_view">
-            <field name="name">res.company.form.inherit.bvr</field>
+            <field name="name">res.company.form.inherit.isr</field>
             <field name="model">res.company</field>
             <field name="inherit_id" ref="base.view_company_form"/>
```
Comment (grindtildeath) [inline l10n_ch_payment_slip/views/company.xml]: IMO would be nice to move these settings from the company from view to res.config.settings

Edit : Didn't see you todo ;)

File: l10n_ch_payment_slip/__manifest__.py [hunk 6]
```diff
@@ -1,32 +1,28 @@
...
  'category': 'Localization',
  'website': 'http://www.camptocamp.com',
  'license': 'AGPL-3',
  'depends': [
-     'base',
-     'account',
-     'report',
+     'account_invoicing',
      'l10n_ch_base_bank',
      'base_transaction_id',  # OCA/bank-statement-reconcile
```
Comment (grindtildeath) [inline l10n_ch_payment_slip/__manifest__.py]: OCA/account-reconcile

File: l10n_ch_payment_slip/tests/test_payment_slip.py [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
 class TestPaymentSlip(test_common.TransactionCase):
     _compile_get_ref = re.compile(r'[^0-9]')
 
+    def setUp(self):
+        super(TestPaymentSlip, self).setUp()
+        self.report1slip_from_inv = self.env.ref(
+            'l10n_ch_payment_slip.one_slip_per_page_from_invoice',
+        )
+
     def make_bank(self):
```
Comment (grindtildeath) [inline l10n_ch_payment_slip/tests/test_payment_slip.py:26]: Would be great to have these records as demo data as it's impossible to test functionally without it.

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Maybe we need this commit ?
https://github.com/OCA/l10n-switzerland/pull/392/commits/266e82dd019d33f2cf1a948f347bdafe6b8fb614

Also I think we should add a dependency on contacts because that's where the res.bank and res.partner.bank menus are defined, and we need these to configure the module properly.

Moreover I get an error when I try to print ISR : 
```


Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/src/addons/web/controllers/main.py", line 77, in wrap
    return f(*args, **kwargs)
  File "/opt/odoo/src/addons/web/controllers/main.py", line 1527, in index
    action["report_name"], report_ids, report_data, context])
  File "/opt/odoo/src/odoo/http.py", line 115, in dispatch_rpc
    result = dispatch(method, params)
UnboundLocalError: local variable 'dispatch' referenced before assignment
```

File: l10n_ch_payment_slip/models/company.py [hunk 8]
```diff
@@ -1,62 +1,71 @@
...
     _inherit = "res.company"
 
-    bvr_delta_horz = fields.Float(
-        'BVR Horz. Delta (inch)',
-        help='horiz. delta in inch 1.2 will print the bvr 1.2 inch lefter, '
-        'negative value is possible'
+    isr_delta_horz = fields.Float(
+        'ISR Horz. Delta (inch)',
+        oldname='bvr_delta_horz',
+        help='horiz. delta in inch 1.2 will print the ISR 1.2 inch lefter,'
```
Comment (tschanzt) [inline l10n_ch_payment_slip/models/company.py]: lefter sounds weird. to the left?

**Review by tschanzt — COMMENTED** [review]
👍 if you get travis to run.

**Comment by yvaucher** [comment]
I'll close this PR as odoo is now managing the payment slip officially again.

If the official module lacks features we can still create other modules. On the other hand it might not be the right time to make lot of changes as the QR invoices will replace this.

**Comment by yvaucher** [comment]
Reopened as official module is currently not providing all required features. (alignment on printout, reconciliation based on BVR reference)

**Comment by grindtildeath** [comment]
@TDu Can you please integrate the content of this PR as well ? https://github.com/OCA/l10n-switzerland/pull/411

**Comment by TDu** [comment]
@grindtildeath It has been merged in 10.0, are you asking to add it to 11.0 ?

**Comment by TDu** [comment]
@grindtildeath Will look at it when I get a green Travis.

**Comment by TDu** [comment]
@grindtildeath Unfortunately it is not straightforward to incorporate your PR in 11.0.  Some methods have moved or been dropped, could you have a look at it ?

**Comment by grindtildeath** [comment]
@TDu I cherry-picked and tried to fix but there's a few things not working yet, can you please check it as you already did most of the migration ? Thx
https://github.com/yvaucher/l10n-switzerland/pull/2

**Comment by yvaucher** [comment]
@grindtildeath thanks I added your changes in here.

**Review by simahawk — APPROVED** [review]

**Comment by simahawk** [comment]
reminder: squash before merging ;)

File: l10n_ch_payment_slip/report/ir_action.py [hunk 9]
```diff
@@ -1,46 +1,201 @@
...
+                    save_in_attachment['loaded_documents'].get(
+                        doc.invoice_id.id)
+            ):
+                with closing(os.fdopen(pdfreport_fd, 'w')) as pdfreport:
+                    pdfreport.write(save_in_attachment['loaded_documents']
+                                    [doc.invoice_id.id])
+                pdfdocuments.append(pdfreport_path)
+                continue
+
+            with closing(os.fdopen(pdfreport_fd, 'w')) as pdfreport:
```
Comment (ghost) [inline l10n_ch_payment_slip/report/ir_action.py]: @yvaucher can you replace `'w'` to `'wb'` or should I create PR in your repo?

File: l10n_ch_payment_slip/report/ir_action.py [hunk 10]
```diff
@@ -1,46 +1,201 @@
...
+    def render_qweb_pdf(self, res_ids=None, data=None):
+        if (self.name == 'l10n_ch_payment_slip.'
+                         'one_slip_per_page_from_invoice'):
+            report = self._get_report_from_name(self.name)
+            save_in_attachment = self._check_attachment_use(
+                docids, report)
+
+            reports = self.browse(docids)
+            return reports._generate_one_slip_per_page_from_invoice_pdf(
+                report_name=report_name, save_in_attachment=save_in_attachment
```
Comment (yvaucher) [inline l10n_ch_payment_slip/report/ir_action.py]: report_name is undefined

File: l10n_ch_payment_slip/report/ir_action.py [hunk 11]
```diff
@@ -1,46 +1,201 @@
...
+
+    @api.multi
+    def render_qweb_pdf(self, res_ids=None, data=None):
+        if (self.name == 'l10n_ch_payment_slip.'
+                         'one_slip_per_page_from_invoice'):
+            report = self._get_report_from_name(self.name)
+            save_in_attachment = self._check_attachment_use(
+                docids, report)
+
+            reports = self.browse(docids)
```
Comment (yvaucher) [inline l10n_ch_payment_slip/report/ir_action.py]: docids is undefined

File: l10n_ch_payment_slip/report/ir_action.py [hunk 12]
```diff
@@ -1,46 +1,201 @@
...
+                limit=1
+            )
+        return report
+
+    @api.multi
+    def render_qweb_pdf(self, res_ids=None, data=None):
+        if (self.name == 'l10n_ch_payment_slip.'
+                         'one_slip_per_page_from_invoice'):
+            report = self._get_report_from_name(self.name)
+            save_in_attachment = self._check_attachment_use(
```
Comment (yvaucher) [inline l10n_ch_payment_slip/report/ir_action.py]: _check_attachment_use does not exist anymore in Odoo 11.0.

This was removed here 

https://github.com/odoo/odoo/commit/e80238042c9d93d492ea8b06b0041aced0d81dcd#diff-c941fd9fa97231fc06b29633532dd307L318

**Comment by BT-ojossen** [comment]
hi, what's the state here? when will this be merged?

**Comment by yvaucher** [comment]
@BT-ojossen this is in a ready state, but for the unit tests which are still red. Thus to answer your question, it will be merged when the tests are fixed.

File: l10n_ch_payment_slip/wizard/isr_batch_print.py [hunk 13]
```diff
@@ -1,22 +1,21 @@
...
 
-class BvrBatchPrintWizard(models.TransientModel):
+class ISRBatchPrintWizard(models.TransientModel):
 
-    _name = 'bvr.batch.print.wizard'
+    _name = 'isr.batch.print.wizard'
 
     invoice_ids = fields.Many2many(comodel_name='account.invoice',
                                    string='Invoices')
     error_message = fields.Text('Errors')
```
Comment (ghost) [inline l10n_ch_payment_slip/wizard/isr_batch_print.py]: Hi @yvaucher, I think it is a silly mistake but,

When you get the related error in the batch print wizard, the error is editable, in my point of view it is a little bit annoying.

Regards!

Comment (yvaucher) [reply l10n_ch_payment_slip/wizard/isr_batch_print.py]: @BT-ccebrian makes sense to me.

@grindtildeath can you also add a readonly on that field? AFAIK it should only be set server side.

Comment (grindtildeath) [reply l10n_ch_payment_slip/wizard/isr_batch_print.py]: Fixed in dee27ae205486dfc0b8e6f217a657567086ed1e3

**Review by simahawk — APPROVED** [review]
LG.
Runbot is not really red, is yellow because of warnings unrelated to this module.

One thing: we might want to mention in a TODO that the merge mode machinery could/should be splitted into reporting-engine module(s)

**Review by tschanzt — APPROVED** [review]

**Comment by yvaucher** [comment]
@grindtildeath LG thanks for fixing the tests

**Review by grindtildeath — APPROVED** [review]
I approve as I'm the author of the last commits :blush:

**Pull request merged**